tutorials/multi/2_node_classification.py

```diff
@@ -0,0 +1,247 @@
+train_nids = idx_split['train']
+valid_nids = idx_split['valid']
+test_nids = idx_split['test']
+
+
+######################################################################
+# Defining Model
+# --------------
+# 
+# The model will be again identical to the :doc:`Training GNN with Neighbor
+# Sampling for Node Classification <../large/L1_large_node_classification>`
+# tutorial.
+# 
+
+class Model(nn.Module):
+    def __init__(self, in_feats, h_feats, num_classes):
+        super(Model, self).__init__()
+        self.conv1 = SAGEConv(in_feats, h_feats, aggregator_type='mean')
+        self.conv2 = SAGEConv(h_feats, num_classes, aggregator_type='mean')
+        self.h_feats = h_feats
+
+    def forward(self, mfgs, x):
+        h_dst = x[:mfgs[0].num_dst_nodes()]
+        h = self.conv1(mfgs[0], (x, h_dst))
+        h = F.relu(h)
+        h_dst = h[:mfgs[1].num_dst_nodes()]
+        h = self.conv2(mfgs[1], (h, h_dst))
+        return h
+
+
+######################################################################
+# Defining Training Procedure
+# ---------------------------
+# 
+# The training procedure will be slightly different from what you saw
+# previously, in the sense that you will need to \* Initialize a
+# distributed training context with ``torch.distributed``. \* Wrap your
+# model with ``torch.nn.parallel.DistributedDataParallel``. \* Add a
+# ``use_ddp=True`` argument to the DGL dataloader you wish to run together
+# with DDP.
+# 
+# You will also need to wrap the training loop inside a function so that
+# you can spawn subprocesses to run it.
+# 
+
+def run(proc_id, devices):
+    # Initialize distributed training context.
+    dev_id = devices[proc_id]
+    dist_init_method = 'tcp://{master_ip}:{master_port}'.format(master_ip='127.0.0.1', master_port='12345')
+    # To use Multi-GPU, change 'gloo' to 'nccl'.
+    torch.distributed.init_process_group(
+        backend='gloo', init_method=dist_init_method, world_size=len(devices), rank=proc_id)
+    device = torch.device('cpu')
+    # Uncomment the following two lines to use GPU.
+    #torch.cuda.set_device(dev_id)
+    #device = torch.device('cuda:' + str(dev_id))
+    
+    # Define training and validation dataloader, copied from the previous tutorial
+    # but with one line of difference: use_ddp to enable distributed data parallel
+    # data loading.
+    sampler = dgl.dataloading.MultiLayerNeighborSampler([4, 4])
+    train_dataloader = dgl.dataloading.NodeDataLoader(
+        # The following arguments are specific to NodeDataLoader.
+        graph,              # The graph
+        train_nids,         # The node IDs to iterate over in minibatches
+        sampler,            # The neighbor sampler
+        device=device,      # Put the sampled MFGs on CPU or GPU
+        use_ddp=True,       # Make it work with distributed data parallel
+        # The following arguments are inherited from PyTorch DataLoader.
+        batch_size=1024,    # Batch size
```

Comment:
Is it per-device batch_size? Maybe we need to address that the total effective batch size is this `batch_size * num_devices`.

Reply:
Done.